Specialize division and remainder of unsigned using strength_reduce

ritchie46 opened the pull request:
Already make this PR to start the discussion. I implemented the strength reduction algorithm discussed in #259 for division of `u64` arrays and the performance increases are almost 10x!

I benchmarked a divsion by 4 (which is a single bit shift) and a divison by a prime number.

```
Gnuplot not found, using plotters backend
divide_scalar 4         time:   [41.871 us 41.883 us 41.899 us]                             
                        change: [-89.393% -89.388% -89.382%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) high mild
  9 (9.00%) high severe

divide_scalar prime     time:   [71.806 us 71.821 us 71.845 us]                                
                        change: [-82.082% -82.035% -81.993%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  5 (5.00%) high severe
```

@jorgecarleitao, I used some `unsafe` to convince the compiler of the proper types. How do you want to go about this. We could also create some extra traits to do the conversion.

jorgecarleitao commented on src/compute/arithmetics/basic/div.rs:
```diff
@@ -109,10 +111,28 @@ where
+            //  Safety:
+            // We need to go from generic to concrete type and back to generic.
+            // We know the types are the same, but the compiler needs some convincing
+            unsafe {
```
remove 1 `unsafe`:

```
           let lhs = (lhs as &dyn Array)
                .as_any()
                .downcast_ref::<PrimitiveArray<u64>>()
                .unwrap();
            let rhs = rhs.to_u64().unwrap();
            let reduced_div = StrengthReducedU64::new(rhs);
            let result = unary(lhs, |a| a / reduced_div, lhs.data_type().clone());
            // Safety: we just proved that `lhs` is `PrimitiveArray<u64>` which means that
            // T = u64
            unsafe { std::mem::transmute::<PrimitiveArray<u64>, PrimitiveArray<T>>(result) }
```

ritchie46 replied:
:+1:

Dandandan replied:
Does this mean the "transmute-free" claim must go?

ritchie46 replied:
Transmute-lean should be the goal😁

jorgecarleitao replied:
at least when I wrote "transmute-free", I was thinking about `arrow-rs` treating every buffer as `&[u8]` and converting them to `&[T]` without being able to prove (only trust) that the buffers are indeed created from `&[T]` (and valid for some size N).

We do use transmutes in the crate atm; e.g. the IPC has [one transmute](https://github.com/jorgecarleitao/arrow2/blob/main/src/io/ipc/read/read_basic.rs#L69) when reading bytes to `&mut [T]`, because there is no way of doing this in safe Rust (despite being `safe`).

imo this is another of those cases: the way to achieve this in safe rust is via trait specialization, which is only available in the unstable channel.

The relevant aspect imo is that we are in control of the call and know exactly why it is safe.

jorgecarleitao replied:
But maybe there is a way to achieve the same without transmutes; I would definitely prefer that!

jorgecarleitao commented on Cargo.toml:
```diff
@@ -60,6 +60,9 @@ ahash = { version = "0.7", optional = true }
 parquet2 = { version = "0.3", optional = true, default_features = false, features = ["stream"] }
 
+# for division/remainder optimization at runtime
+strength_reduce = "0.2"
```
optional under feature `compute`?

ritchie46 replied:
:+1:

jorgecarleitao commented on benches/arithmetic_kernels.rs:
```diff
@@ -0,0 +1,53 @@
+fn add_benchmark(c: &mut Criterion) {
+    let size = 65536;
+    let arr = create_primitive_array_with_seed::<u64>(size, DataType::UInt64, 0.0, 43);
+    let arr_i64 = create_primitive_array_with_seed::<i64>(size, DataType::Int64, 0.0, 43);
```
not used.

ritchie46 replied:
Ah.. forgot about that one.

jorgecarleitao reviewed (commented):
Overall looks great!

I think that for now we should live with the transmute. I think that the new scalar API will enable us to write this function as `div_scalar(&dyn Array, &dyn Scalar) -> Box<dyn Array>`, which will allow us to avoid this transmute and specialize directly based on `DataType`.

u16 and u32 is also supported by strength division.

ritchie46 commented:
> I think that for now we should live with the transmute. I think that the new scalar API will enable us to write this function as div_scalar(&dyn Array, &dyn Scalar) -> Box<dyn Array>, which will allow us to avoid this transmute and specialize directly based on DataType.

Great! than I will implement the other types as well.

ritchie46 commented:
@jorgecarleitao I updated the PR. It is now implemented for `rem_scalar` and `div_scalar`. 

There is a bit repetition in the code. Because branches work on different concrete types it seems not possible to write a generic function for that. A macro would be possible but, but given the usage of `unsafe`, I chose for repetition as that's more readable.

jorgecarleitao commented:
Do we have a test hitting the branch?

ritchie46 commented:
> Do we have a test hitting the branch?

Will write that tomorrow.

ritchie46 commented:
@jorgecarleitao all branches are rain in tests now. I think its good to go.
